File: gin/library/curl.go
package library

import (
	"bytes"
	"io/ioutil"
	"net/http"
	"net/url"
	"strings"
	"webmis/util"
)

/* 请求 */
type Curl struct{}

/* GET、POST、PUT、HEAD、DELETE */
func (c Curl) Request(url string, data []byte, method string, header map[string]interface{}) ([]byte, error) {
	// 请求头
	param := map[string]interface{}{
		"Content-Type": "application/json; charset=utf-8", //JSON方式
	}
	if header != nil {
		param = util.ArrayMerge(param, header)
	}
	// 发送
	req, err := http.NewRequest(method, url, bytes.NewBuffer(data))
	if err != nil {
		return nil, err
	}
	for k, v := range param {
		req.Header.Set(k, (&util.Type{}).Strval(v))
	}
	resp, err := (&http.Client{}).Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	return ioutil.ReadAll(resp.Body)
}

/* URL参数-生成 */
func (c Curl) UrlEncode(data map[string]interface{}) string {
	res := ""
	for k, v := range data {
		res += k + "=" + url.QueryEscape((&util.Type{}).Strval(v)) + "&"
	}
	res = strings.TrimRight(res, "&")
	return res
}

/* URL参数-解析 */
func (c Curl) UrlDecode(data string) map[string]interface{} {
	res := map[string]interface{}{}
	arr := strings.Split(data, "&")
	for _, v := range arr {
		tmp := strings.Split(v, "=")
		if len(tmp) == 2 {
			res[tmp[0]], _ = url.QueryUnescape(tmp[1])
		}
	}
	return res
}

File: phalcon/library/Curl.php
<?php
namespace Library;

use Service\Base;

/* 请求 */
class Curl extends Base {

  /* GET、POST、PUT、HEAD、DELETE */
  static function Request(string $url, string $data='', string $method='GET', array $headers=[], string $resType='json') {
    // 请求头
    $headerArr = [];
    foreach($headers as $k=>$v){
      $headerArr[] = $k.': '.$v;
    }
    // 发送
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
    curl_setopt($ch,CURLOPT_HTTPHEADER,$headerArr);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    // 数据
    if($method!='GET'){
      curl_setopt($ch, CURLOPT_POSTFIELDS, $data);
    }
    $text = curl_exec($ch);
    curl_close($ch);
    // 结果
    if($resType=='json') $res = !empty($text)?json_decode($text):null;
    else $res = $text;
    return $res;
  }

  /* URL参数-生成 */
  static function UrlEncode(array $data): string {
    return http_build_query($data);
  }
  /* URL参数-解析 */
  static function UrlDecode(string $data): array {
    $res = [];
    $arr = explode('&', $data);
    foreach($arr as $v){
      $tmp = explode('=', $v);
      if($tmp[1]) $res[$tmp[0]]=urldecode($tmp[1]);
    }
    return $res;
  }

}